spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/mapping/BasicCassandraPersistentEntity.java

```diff
@@ -226,6 +226,12 @@ public CqlIdentifier getTableName() {
 		return Optional.ofNullable(this.tableName).orElseGet(this::determineTableName);
 	}
 
+	@Nullable
+	@Override
+	public CqlIdentifier getKeyspace() {
```

Comment:
Let's turn that method into an `Optional`. We should make use on the metadata-level of improved null-safety.

Reply:
The question is at which level we should transform this `Optional` back to a value.
I.e: when following your suggestion I need to change all callers and this will be propagated up to specific method, i.e:
```
public StatementBuilder<Select> count(Query query, CassandraPersistentEntity<?> entity,
           Optional<CqlIdentifier> keyspace, CqlIdentifier tableName) //
``` 
but it will generate the `Optional<CqlIdentifier> used as a type for parameter keyspace` warning; this is the reason why I did use `@Nullable` instead of an `Optional`. 
Are we ok with a warning?
If not, but If we still want to change those methods to return `Optional` (`CassandraPersistemtEntity#getKeyspace` AND `BasicCassandraPersistemtEntity#getKeyspace`) then we need to transform from `Optional` to a value in every caller, i.e:
```
public StatementBuilder<Select> count(Query query, CassandraPersistentEntity<?> persistentEntity) {
               // ....
               return count(query, persistentEntity, persistentEntity.getKeyspace().orElse(null), persistentEntity.getTableName());
	}
```
and this `persistentEntity.getKeyspace().orElse(null)` will be propagated in a lot of places in code so that is far from ideal.

Reply:
> The question is at which level we should transform this Optional back to a value.

Ideally, all methods accepting `CqlIdentifier` as table name get overloaded (if they are public) to accept a `TableName`/`TableCoordinates`/`CqlTableName` object so that keyspace retrieval happens as late as possible and in as little places as possible.

Something along the lines of:


```java
private static StatementBuilder<Select> createSelectAndOrder(List<Selector> selectors, CqlTableName from,
		Filter filter, Sort sort) {

	Select select;

	if (selectors.isEmpty()) {
		select = from.map((keyspace, tableName) -> QueryBuilder.selectFrom(keyspace, tableName)).orElseGet(tableName -> QueryBuilder.selectFrom(tableName)).all();
```

which was previously:

```java
private static StatementBuilder<Select> createSelectAndOrder(List<Selector> selectors, CqlIdentifier from,
		Filter filter, Sort sort) {

	Select select;

	if (selectors.isEmpty()) {
		select = QueryBuilder.selectFrom(from).all();
```

As mentioned, I'm not 100% convinced about naming the value object `TableName`, so happy to discuss suggestions.

On the `PersistentEntity`-level it makes sense to keep getters for `keyspaceName` and `tableName`.

Reply:
Regarding the naming, I think that it would be beneficial to have the `Keysapce` in the name of this new type.
So maybe just `KeyspaceTable` OR  `KeyspaceTableName`?

Reply:
Rather than enumerating the properties of which the information consists, we should name the thing that it describes.

Reply:
In that case, I think that `TableCoordinates` describes it best.

Reply:
Sounds good. Then we can start working on that one as value object that consists of the actual name and an optional keyspace name. Typically, we hide constructors and provide factory methods along the lines of `of(CqlIdentifier tableName, Optional<CqlIdentifier> keyspaceName)`. We also should consider how the contents gets used. Besiders getters it would make sense to allow a somewhat functional approach that allows mapping of both information (if present) and fall back to the table name if the keyspace isn't specified.

I'll be off for about two weeks, so feel free to experiment with what feels best.

Reply:
regarding mapping, I think that this is not necessary. All `queryBuilder` methods are accepting null keyspace (and ignore it if it's null):
```
@NonNull
  public static InsertInto insertInto(
      @Nullable CqlIdentifier keyspace, @NonNull CqlIdentifier table) {
```
so we can just pass `null`:
```
QueryBuilder.insertInto(tableCoordinates.getKeyspaceName().orElse(null), tableCoordinates.getTableName())
```
without a need for `optional.map`.